File: roc_module_lua/Lua/LuaDefs/QuaternionDefs.cpp
#include "stdafx.h"
#include "Lua/LuaDefs/QuaternionDefs.h"
#include "Lua/LuaDefs.h"
#include "Lua/LuaVM.h"
#include "Lua/LuaArgReader.h"
#include "Utils.h"

extern const std::string g_vec3Name;
extern const std::string g_quatName("Quaternion");

std::vector<LuaPropDef> QuaternionDefs::ms_staticProps;
std::vector<LuaMethodDef> QuaternionDefs::ms_staticMethods;
std::vector<LuaMethodDef> QuaternionDefs::ms_metaMethods;
std::vector<LuaPropDef> QuaternionDefs::ms_instanceProps;
std::vector<LuaMethodDef> QuaternionDefs::ms_instanceMethods;

void QuaternionDefs::Init()
{
    ms_staticProps.emplace_back("identity", GetIdentity, nullptr);

    ms_staticMethods.emplace_back("euler", Euler);

    ms_metaMethods.emplace_back("__mul", Multiply);

    ms_instanceProps.emplace_back("x", GetX, SetX);
    ms_instanceProps.emplace_back("y", GetY, SetY);
    ms_instanceProps.emplace_back("z", GetZ, SetZ);
    ms_instanceProps.emplace_back("w", GetW, SetW);

    ms_instanceMethods.emplace_back("rotate", Rotate);
}

void QuaternionDefs::RegisterInVM(LuaVM *p_vm)
{
    p_vm->RegisterLuaClass(g_quatName, Create, &ms_staticProps, &ms_staticMethods, &ms_metaMethods, &ms_instanceProps, &ms_instanceMethods);
    p_vm->RegisterFunction("isQuaternion", IsQuaternion);
}

int QuaternionDefs::Create(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    float l_x = 0.f;
    float l_y = 0.f;
    float l_z = 0.f;
    float l_w = 0.f;
    l_argReader.Skip();
    l_argReader.ReadNextNumber(l_x);
    l_argReader.ReadNextNumber(l_y);
    l_argReader.ReadNextNumber(l_z);
    l_argReader.ReadNextNumber(l_w);
    l_argReader.PushObject(new glm::quat(l_w, l_x, l_y, l_z), g_quatName, false);
    return l_argReader.GetReturnValue();
}

int QuaternionDefs::GetIdentity(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    l_argReader.PushObject(new glm::quat(1.f, 0.f, 0.f, 0.f), g_quatName, false);
    return 1;
}

int QuaternionDefs::Euler(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::vec3 *l_vec;
    l_argReader.ReadObject(l_vec, g_vec3Name);
    if(!l_argReader.HasError())
        l_argReader.PushObject(new glm::quat(*l_vec), g_quatName, false);
    else
            l_argReader.PushBoolean(false);

    l_argReader.LogError();
    return l_argReader.GetReturnValue();
}

int QuaternionDefs::Multiply(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quatA;
    glm::quat *l_quatB;
    l_argReader.ReadObject(l_quatA, g_quatName);
    l_argReader.ReadObject(l_quatB, g_quatName);
    if(!l_argReader.HasError())
        l_argReader.PushObject(new glm::quat(*l_quatA * *l_quatB), g_quatName, false);
    else
        l_argReader.PushBoolean(false);

    l_argReader.LogError();
    return 1;
}

int QuaternionDefs::GetX(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat;
    l_argReader.ReadObject(l_quat, g_quatName);
    if(!l_argReader.HasError())
        l_argReader.PushNumber(l_quat->x);
    else
        l_argReader.PushBoolean(false);

    l_argReader.LogError();
    return 1;
}
int QuaternionDefs::SetX(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat;
    float l_value;
    l_argReader.ReadObject(l_quat, g_quatName);
    l_argReader.ReadNumber(l_value);
    if(!l_argReader.HasError())
        l_quat->x = l_value;

    l_argReader.LogError();
    return 0;
}

int QuaternionDefs::GetY(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat;
    l_argReader.ReadObject(l_quat, g_quatName);
    if(!l_argReader.HasError())
        l_argReader.PushNumber(l_quat->y);
    else
        l_argReader.PushBoolean(false);

    l_argReader.LogError();
    return 1;
}
int QuaternionDefs::SetY(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat;
    float l_value;
    l_argReader.ReadObject(l_quat, g_quatName);
    l_argReader.ReadNumber(l_value);
    if(!l_argReader.HasError())
        l_quat->y = l_value;

    l_argReader.LogError();
    return 0;
}

int QuaternionDefs::GetZ(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat;
    l_argReader.ReadObject(l_quat, g_quatName);
    if(!l_argReader.HasError())
        l_argReader.PushNumber(l_quat->z);
    else
        l_argReader.PushBoolean(false);

    l_argReader.LogError();
    return 1;
}
int QuaternionDefs::SetZ(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat;
    float l_value;
    l_argReader.ReadObject(l_quat, g_quatName);
    l_argReader.ReadNumber(l_value);
    if(!l_argReader.HasError())
        l_quat->z = l_value;

    l_argReader.LogError();
    return 0;
}

int QuaternionDefs::GetW(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat;
    l_argReader.ReadObject(l_quat, g_quatName);
    if(!l_argReader.HasError())
        l_argReader.PushNumber(l_quat->w);
    else
        l_argReader.PushBoolean(false);

    l_argReader.LogError();
    return 1;
}
int QuaternionDefs::SetW(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat;
    float l_value;
    l_argReader.ReadObject(l_quat, g_quatName);
    l_argReader.ReadNumber(l_value);
    if(!l_argReader.HasError())
        l_quat->w = l_value;

    l_argReader.LogError();
    return 0;
}

int QuaternionDefs::Rotate(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat;
    glm::vec3 *l_vec;
    l_argReader.ReadObject(l_quat, g_quatName);
    l_argReader.ReadObject(l_vec, g_vec3Name);
    if(!l_argReader.HasError())
        l_argReader.PushObject(new glm::vec3(*l_quat * *l_vec), g_vec3Name, false);
    else
        l_argReader.PushBoolean(false);

    l_argReader.LogError();
    return l_argReader.GetReturnValue();
}

int QuaternionDefs::IsQuaternion(lua_State *p_state)
{
    LuaArgReader l_argReader(p_state);
    glm::quat *l_quat = nullptr;
    l_argReader.ReadNextObject(l_quat, g_quatName);
    l_argReader.PushBoolean(l_quat != nullptr);
    return l_argReader.GetReturnValue();
}

File: roc_module_lua/Defs/QuaternionDefs.cs
using GlmSharp;
using LuaSharp.Lua;
using ROC.Module.Wrappers;
using System;
using System.Collections.Generic;

namespace ROC.Module.Defs
{
    internal static class QuaternionDefs
    {
        public static readonly LuaVM.LuaClassDefinition Definition = new LuaVM.LuaClassDefinition();
        static readonly Type ms_quaternionType = typeof(Quaternion);
        static readonly Type ms_vector3Type = typeof(Vector3);
        static readonly Type ms_vector4Type = typeof(Vector4);

        static QuaternionDefs()
        {
            Definition.m_name = nameof(Quaternion);
            Definition.m_constructor = Create;

            Definition.m_staticProperties = new List<(string, LuaInterop.lua_CFunction, LuaInterop.lua_CFunction)>()
            {
                ("identity", CreateIdentity, null)
            };

            Definition.m_staticMethods = new List<(string, LuaInterop.lua_CFunction)>()
            {
                ("FromEuler", FromEuler),
                ("FromAxisAngle", FromAxisAngle),
                ("Cross", Cross),
                ("Dot", Dot),
                ("Lerp", Lerp),
                ("Slerp", SLerp)
            };

            Definition.m_metaMethods = new List<(string, LuaInterop.lua_CFunction)>()
            {
                ("__mul", Multiply),
                ("__add", Add),
                ("__sub", Subtract),
                ("__div", Divide)
            };

            Definition.m_instanceProperties = new List<(string, LuaInterop.lua_CFunction, LuaInterop.lua_CFunction)>()
            {
                ("x", GetX,SetX),
                ("y", GetY,SetY),
                ("z", GetZ,SetZ),
                ("w", GetW,SetW),
                ("roll", GetRoll, null),
                ("pitch", GetPitch, null),
                ("yaw", GetYaw, null),
                ("normalized", Normalized, null),
                ("inversed", Inversed, null),
                ("conjugated", Conjugated, null),
            };
        }

        static int Create(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            l_argReader.Skip();

            if(!l_argReader.ReadNumber(out float l_x) || !l_argReader.ReadNumber(out float l_y) || !l_argReader.ReadNumber(out float l_z) || !l_argReader.ReadNumber(out float l_w))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushObject(new Quaternion(new quat(l_x, l_y, l_z, l_w)), ms_quaternionType);
            return 1;
        }

        static int CreateIdentity(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            l_argReader.PushObject(new Quaternion(quat.Identity), ms_quaternionType);
            return 1;
        }

        static int FromEuler(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(l_argReader.IsNextNumber())
            {
                if(!l_argReader.ReadNumber(out float l_x) || !l_argReader.ReadNumber(out float l_y) || !l_argReader.ReadNumber(out float l_z))
                {
                    l_argReader.PushBoolean(false);
                    return 1;
                }

                l_argReader.PushObject(new Quaternion(new quat(new vec3(l_x, l_y, l_z))), ms_quaternionType);
                return 1;
            }

            if(l_argReader.IsNextObject())
            {
                if(!l_argReader.ReadObject(out Vector3 l_vec))
                {
                    l_argReader.PushBoolean(false);
                    return 1;
                }

                l_argReader.PushObject(new Quaternion(new quat(l_vec.m_vector)), ms_quaternionType);
                return 1;
            }

            l_argReader.PushBoolean(false);
            return 1;
        }

        static int FromAxisAngle(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadNumber(out float l_angle) || !l_argReader.ReadObject(out Vector3 l_axis))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushObject(new Quaternion(quat.FromAxisAngle(l_angle, l_axis.m_vector)), ms_quaternionType);
            return 1;
        }

        static int Cross(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quatA) || !l_argReader.ReadObject(out Quaternion l_quatB))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushObject(new Quaternion(quat.Cross(l_quatA.m_quat, l_quatB.m_quat)), ms_quaternionType);
            return 1;
        }

        static int Dot(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quatA) || !l_argReader.ReadObject(out Quaternion l_quatB))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushNumber(quat.Dot(l_quatA.m_quat, l_quatB.m_quat));
            return 1;
        }

        static int Lerp(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quatA) || !l_argReader.ReadObject(out Quaternion l_quatB) || !l_argReader.ReadNumber(out float l_alpha))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushObject(new Quaternion(quat.Lerp(l_quatA.m_quat, l_quatB.m_quat, l_alpha)), ms_quaternionType);
            return 1;
        }

        static int SLerp(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quatA) || !l_argReader.ReadObject(out Quaternion l_quatB) || !l_argReader.ReadNumber(out float l_alpha))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushObject(new Quaternion(quat.SLerp(l_quatA.m_quat, l_quatB.m_quat, l_alpha)), ms_quaternionType);
            return 1;
        }

        static int Multiply(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadValue(out object l_objA) || !l_argReader.ReadValue(out object l_objB))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            // Quaternion * value
            if(l_objA is Quaternion l_quat0 && l_objB is double l_val0)
            {
                l_argReader.PushObject(new Quaternion(l_quat0.m_quat * (float)l_val0), ms_quaternionType);
                return 1;
            }
            if(l_objA is Quaternion l_quat1 && l_objB is long l_val1)
            {
                l_argReader.PushObject(new Quaternion(l_quat1.m_quat * l_val1), ms_quaternionType);
                return 1;
            }
            if(l_objA is Quaternion l_quat2 && l_objB is Quaternion l_val2)
            {
                l_argReader.PushObject(new Quaternion(l_quat2.m_quat * l_val2.m_quat), ms_quaternionType);
                return 1;
            }
            if(l_objA is Quaternion l_quat3 && l_objB is Vector3 l_val3)
            {
                l_argReader.PushObject(new Vector3(l_quat3.m_quat * l_val3.m_vector), ms_vector3Type);
                return 1;
            }
            if(l_objA is Quaternion l_quat4 && l_objB is Vector4 l_val4)
            {
                l_argReader.PushObject(new Vector4(l_quat4.m_quat * l_val4.m_vector), ms_vector4Type);
                return 1;
            }

            // Value * quaternion
            if(l_objA is double l_val5 && l_objB is Quaternion l_quat5)
            {
                l_argReader.PushObject(new Quaternion((float)l_val5 * l_quat5.m_quat), ms_quaternionType);
                return 1;
            }
            if(l_objA is long l_val6 && l_objB is Quaternion l_quat6)
            {
                l_argReader.PushObject(new Quaternion(l_val6 * l_quat6.m_quat), ms_quaternionType);
                return 1;
            }
            if(l_objA is Vector3 l_val7 && l_objB is Quaternion l_quat7)
            {
                l_argReader.PushObject(new Vector3(l_val7.m_vector * l_quat7.m_quat), ms_vector3Type);
                return 1;
            }
            if(l_objA is Vector4 l_val8 && l_objB is Quaternion l_quat8)
            {
                l_argReader.PushObject(new Vector4(l_val8.m_vector * l_quat8.m_quat), ms_vector4Type);
                return 1;
            }

            l_argReader.PushBoolean(false);
            return 1;
        }

        static int Add(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadValue(out object l_objA) || !l_argReader.ReadValue(out object l_objB))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            // Quaternion + value
            if(l_objA is Quaternion l_quat0 && l_objB is double l_val0)
            {
                l_argReader.PushObject(new Quaternion(l_quat0.m_quat + (float)l_val0), ms_quaternionType);
                return 1;
            }
            if(l_objA is Quaternion l_quat1 && l_objB is long l_val1)
            {
                l_argReader.PushObject(new Quaternion(l_quat1.m_quat + l_val1), ms_quaternionType);
                return 1;
            }
            if(l_objA is Quaternion l_quat2 && l_objB is Quaternion l_val2)
            {
                l_argReader.PushObject(new Quaternion(l_quat2.m_quat + l_val2.m_quat), ms_quaternionType);
                return 1;
            }

            // Value + quaternion
            if(l_objA is double l_val5 && l_objB is Quaternion l_quat5)
            {
                l_argReader.PushObject(new Quaternion((float)l_val5 + l_quat5.m_quat), ms_quaternionType);
                return 1;
            }
            if(l_objA is long l_val6 && l_objB is Quaternion l_quat6)
            {
                l_argReader.PushObject(new Quaternion(l_val6 + l_quat6.m_quat), ms_quaternionType);
                return 1;
            }

            l_argReader.PushBoolean(false);
            return 1;
        }

        static int Subtract(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadValue(out object l_objA) || !l_argReader.ReadValue(out object l_objB))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            // Quaternion - value
            if(l_objA is Quaternion l_quat0 && l_objB is double l_val0)
            {
                l_argReader.PushObject(new Quaternion(l_quat0.m_quat - (float)l_val0), ms_quaternionType);
                return 1;
            }
            if(l_objA is Quaternion l_quat1 && l_objB is long l_val1)
            {
                l_argReader.PushObject(new Quaternion(l_quat1.m_quat - l_val1), ms_quaternionType);
                return 1;
            }
            if(l_objA is Quaternion l_quat2 && l_objB is Quaternion l_val2)
            {
                l_argReader.PushObject(new Quaternion(l_quat2.m_quat - l_val2.m_quat), ms_quaternionType);
                return 1;
            }

            // Value - quaternion
            if(l_objA is double l_val5 && l_objB is Quaternion l_quat5)
            {
                l_argReader.PushObject(new Quaternion((float)l_val5 - l_quat5.m_quat), ms_quaternionType);
                return 1;
            }
            if(l_objA is long l_val6 && l_objB is Quaternion l_quat6)
            {
                l_argReader.PushObject(new Quaternion(l_val6 - l_quat6.m_quat), ms_quaternionType);
                return 1;
            }

            l_argReader.PushBoolean(false);
            return 1;
        }

        static int Divide(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadValue(out object l_objA) || !l_argReader.ReadValue(out object l_objB))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            // Quaternion / value
            if(l_objA is Quaternion l_quat0 && l_objB is double l_val0)
            {
                l_argReader.PushObject(new Quaternion(l_quat0.m_quat / (float)l_val0), ms_quaternionType);
                return 1;
            }
            if(l_objA is Quaternion l_quat1 && l_objB is long l_val1)
            {
                l_argReader.PushObject(new Quaternion(l_quat1.m_quat / l_val1), ms_quaternionType);
                return 1;
            }

            l_argReader.PushBoolean(false);
            return 1;
        }

        static int GetX(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushNumber(l_quat.m_quat.x);
            return 1;
        }
        static int SetX(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat) || !l_argReader.ReadNumber(out float l_val))
                return 0;

            l_quat.m_quat.x = l_val;
            return 0;
        }

        static int GetY(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushNumber(l_quat.m_quat.y);
            return 1;
        }
        static int SetY(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat) || !l_argReader.ReadNumber(out float l_val))
                return 0;

            l_quat.m_quat.y = l_val;
            return 0;
        }

        static int GetZ(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushNumber(l_quat.m_quat.z);
            return 1;
        }
        static int SetZ(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat) || !l_argReader.ReadNumber(out float l_val))
                return 0;

            l_quat.m_quat.z = l_val;
            return 0;
        }

        static int GetW(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushNumber(l_quat.m_quat.w);
            return 1;
        }
        static int SetW(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat) || !l_argReader.ReadNumber(out float l_val))
                return 0;

            l_quat.m_quat.w = l_val;
            return 0;
        }

        static int GetPitch(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushNumber(l_quat.m_quat.Pitch);
            return 1;
        }

        static int GetRoll(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushNumber(l_quat.m_quat.Roll);
            return 1;
        }

        static int GetYaw(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushNumber(l_quat.m_quat.Yaw);
            return 1;
        }

        static int Normalized(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushObject(new Quaternion(l_quat.m_quat.Normalized), ms_quaternionType);
            return 1;
        }

        static int Inversed(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushObject(new Quaternion(l_quat.m_quat.Inverse), ms_quaternionType);
            return 1;
        }

        static int Conjugated(IntPtr p_state)
        {
            var l_argReader = new ArgReader(p_state);
            if(!l_argReader.ReadObject(out Quaternion l_quat))
            {
                l_argReader.PushBoolean(false);
                return 1;
            }

            l_argReader.PushObject(new Quaternion(l_quat.m_quat.Conjugate), ms_quaternionType);
            return 1;
        }
    }
}
